### PR TITLE
fix(kumactl) warn version unknown

### PR DESCRIPTION
**Pull request by parkanzky**
### Summary

Added warning in case where server version cannot be retrieved.

This required to update every kumactl test, since they all depended upon there being no kuma-cp on localhost and the resulting connection-refused from the server, on every command, being ignored.

So, added a mock client which always returns a given version, or defaults to current version if none is passed in.

### Full changelog

* Fix silent ignore of missing server version

### Issues resolved

https://github.com/kumahq/kuma/issues/2400

### Documentation

- [ ] Link to the website [documentation PR](https://github.com/kumahq/kuma-website/pull/XXX)

### Testing

- [x] Unit tests
- [ ] E2E tests
- [x] Manual testing on Universal
- [x] Manual testing on Kubernetes 

### Backwards compatibility

- [x] Add `backport-to-stable` label if the code is backwards compatible. Otherwise, list breaking changes.

### CODE REVIEW

**Review by jpeach — COMMENTED** [review]
Does this PR update #2400?

File: app/kumactl/cmd/root.go [hunk 1]
```diff
@@ -74,10 +74,15 @@ func NewRootCmd(root *kumactl_cmd.RootContext) *cobra.Command {
...
+				kumaBuildVersion, err = client.GetVersion(context.Background())
+				if err != nil {
+					kumactlLog.Error(err, "Unable to retrieve server version")
+				}
 			}
 
-			if kumaBuildVersion != nil && (kumaBuildVersion.Version != kuma_version.Build.Version || kumaBuildVersion.Tagline != kuma_version.Product) {
+			if kumaBuildVersion == nil {
+				cmd.PrintErr("WARNING: Unable to retrieve server version. Cannot confirm kumactl matches server version.\n")
+			} else if kumaBuildVersion != nil && (kumaBuildVersion.Version != kuma_version.Build.Version || kumaBuildVersion.Tagline != kuma_version.Product) {
```
[the remark this answers is not in the document]
Comment (bartsmykla) [reply app/kumactl/cmd/root.go]: addressed for Paul

File: app/kumactl/cmd/root.go [hunk 2]
```diff
@@ -74,10 +74,15 @@ func NewRootCmd(root *kumactl_cmd.RootContext) *cobra.Command {
...
-				kumaBuildVersion, _ = client.GetVersion(context.Background())
+				kumaBuildVersion, err = client.GetVersion(context.Background())
+				if err != nil {
+					kumactlLog.Error(err, "Unable to retrieve server version")
+				}
 			}
 
-			if kumaBuildVersion != nil && (kumaBuildVersion.Version != kuma_version.Build.Version || kumaBuildVersion.Tagline != kuma_version.Product) {
+			if kumaBuildVersion == nil {
+				cmd.PrintErr("WARNING: Unable to retrieve server version. Cannot confirm kumactl matches server version.\n")
```
[the remark this answers is not in the document]
Comment (bartsmykla) [reply app/kumactl/cmd/root.go]: addressed for Paul

File: pkg/util/test/mock_api_client.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
+package test
+
+import (
+	"context"
+
+	"github.com/spf13/cobra"
+
+	kumactl_resources "github.com/kumahq/kuma/app/kumactl/pkg/resources"
+
```
[the remark this answers is not in the document]
Comment (bartsmykla) [reply pkg/util/test/mock_api_client.go]: addressed for Paul

File: pkg/util/test/mock_api_client.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+type MockAPIServerClient struct {
+	Version types.IndexResponse
+}
+
+func (m *MockAPIServerClient) GetVersion(ctx context.Context) (*types.IndexResponse, error) {
+	return &m.Version, nil
+}
+
+func GetMockNewAPIServerClient(version *types.IndexResponse) func(*config_proto.ControlPlaneCoordinates_ApiServer) (kumactl_resources.ApiServerClient, error) {
```
[the remark this answers is not in the document]
Comment (bartsmykla) [reply pkg/util/test/mock_api_client.go]: addressed for Paul

File: app/kumactl/cmd/apply/apply_test.go [hunk 5]
```diff
@@ -31,6 +29,7 @@ import (
 	memory_resources "github.com/kumahq/kuma/pkg/plugins/resources/memory"
 	"github.com/kumahq/kuma/pkg/test/resources/model"
 	test_store "github.com/kumahq/kuma/pkg/test/store"
+	kuma_test "github.com/kumahq/kuma/pkg/util/test"
```
[the remark this answers is not in the document]
Comment (bartsmykla) [reply app/kumactl/cmd/apply/apply_test.go]: addressed for Paul

File: pkg/util/test/mock_api_client.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (parkanzky) [reply pkg/util/test/mock_api_client.go]: I was actually planning on adding a test this morning for the server version mismatch and missing warnings, which would have used it..

**Comment by parkanzky** [comment]
> Does this PR update #2400?

Yes it does. Will link.

File: app/kumactl/cmd/root.go [hunk 6]
```diff
@@ -74,10 +74,15 @@ func NewRootCmd(root *kumactl_cmd.RootContext) *cobra.Command {
 			if err != nil {
 				kumactlLog.Error(err, "Unable to get index client")
 			} else {
-				kumaBuildVersion, _ = client.GetVersion(context.Background())
+				kumaBuildVersion, err = client.GetVersion(context.Background())
+				if err != nil {
+					kumactlLog.Error(err, "Unable to retrieve server version")
```
[the remark this answers is not in the document]
Comment (jakubdyszkiewicz) [reply app/kumactl/cmd/root.go:79]: I think we should remove this log altogether. It won't be visible at all. We don't use logging in kumactl to not disrupt stdout.

**Review by jakubdyszkiewicz — APPROVED** [review]

**Pull request merged**